server/src/main/java/org/apache/druid/segment/metadata/SegmentSchemaManager.java

```diff
@@ -0,0 +1,456 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.metadata;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import org.apache.commons.lang.StringEscapeUtils;
+import org.apache.druid.guice.LazySingleton;
+import org.apache.druid.java.util.common.DateTimes;
+import org.apache.druid.java.util.common.ISE;
+import org.apache.druid.java.util.common.Pair;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.emitter.EmittingLogger;
+import org.apache.druid.metadata.MetadataStorageTablesConfig;
+import org.apache.druid.metadata.SQLMetadataConnector;
+import org.apache.druid.segment.SchemaPayload;
+import org.apache.druid.segment.SchemaPayloadPlus;
+import org.apache.druid.timeline.SegmentId;
+import org.skife.jdbi.v2.Handle;
+import org.skife.jdbi.v2.PreparedBatch;
+import org.skife.jdbi.v2.TransactionCallback;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Handles segment schema persistence and cleanup.
+ */
+@LazySingleton
+public class SegmentSchemaManager
+{
+  private static final EmittingLogger log = new EmittingLogger(SegmentSchemaManager.class);
+  private static final int DB_ACTION_PARTITION_SIZE = 100;
+  private final MetadataStorageTablesConfig dbTables;
+  private final ObjectMapper jsonMapper;
+  private final SQLMetadataConnector connector;
+  private final FingerprintGenerator fingerprintGenerator;
+
+  @Inject
+  public SegmentSchemaManager(
+      MetadataStorageTablesConfig dbTables,
+      ObjectMapper jsonMapper,
+      SQLMetadataConnector connector,
+      FingerprintGenerator fingerprintGenerator
+  )
+  {
+    this.dbTables = dbTables;
+    this.jsonMapper = jsonMapper;
+    this.connector = connector;
+    this.fingerprintGenerator = fingerprintGenerator;
+  }
+
+  public List<Long> identifyReferencedUnusedSchema()
+  {
+    return connector.retryWithHandle(
+        handle ->
+            handle.createQuery(
+                      StringUtils.format(
+                          "SELECT DISTINCT(schema_id) FROM %s WHERE used = true AND schema_id IN (SELECT id FROM %s WHERE used = false)",
```

Comment:
The initial deletion logic didn't delete schema for unused segments, but in the latest refactor we changed that logic. 
Now we delete schema both unreferenced schema and schema referenced by unused segments. If a segment is marked as used at a later point, the backfill job will populate its schema. 
This would probably save us some memory on the Coordinator, as we don't have to store schema for unused segments. 
However, the downside is that we loose schema which might be required at a later point.

Reply:
IMHO we should look at only segments which have used=true because: 
* Its a faster query for Big clusters :). 
* Marking a segments from unused to used is a admin initiated step. If the schemas of those segments need to be populated we anyway fire segment metadata q's so its no different from today. The schema persisting can happen with a background thread. 

I personally know of some clusters which have like 10K used segments and 900K unused segments :) 
Let me know if this makes sense cc @kfaraz .

Reply:
Thanks for the clarification, @cryptoe, it makes sense.